apps/prairielearn/src/lib/editors.ts

```diff
@@ -889,16 +925,53 @@ export class CourseInstanceAddEditor extends Editor {
     );
 
     debug('Generate short_name and long_name');
-    const names = this.getNamesForAdd(oldNamesShort, oldNamesLong);
+    const names = this.getNamesForAdd(oldNamesShort, oldNamesLong, this.short_name, this.long_name);
+
     const short_name = names.shortName;
     const courseInstancePath = path.join(courseInstancesPath, short_name);
 
     debug('Write infoCourseInstance.json');
 
+    let allowAccess: { startDate?: string; endDate?: string } | undefined = undefined;
+
+    let startDate: Temporal.ZonedDateTime | undefined;
+    let endDate: Temporal.ZonedDateTime | undefined;
+
+    if (this.start_access_date) {
+      startDate = Temporal.PlainDateTime.from(this.start_access_date).toZonedDateTime(
+        this.course.display_timezone,
+      );
+    }
+
+    if (this.end_access_date) {
+      endDate = Temporal.PlainDateTime.from(this.end_access_date).toZonedDateTime(
+        this.course.display_timezone,
+      );
+    }
+
+    if (startDate && endDate && startDate.epochMilliseconds > endDate.epochMilliseconds) {
+      throw new HttpStatusError(400, 'Start date must be before end date');
+    }
```

Comment:
I'd prefer to push this parsing and validation higher up in the stack. Once we start the edit/sync process, we really don't want it to fail. That is: I think `start_access_date`/`end_access_date` as received in the `params` here should themselves be `Date` or maybe some type of `Temporal` instance.

Reply:
I modified `CourseInstanceAddEditor` to receive `start_access_date` and `end_access_date` as `Temporal.ZonedDateTime` objects, and moved their validation to the constructor:

![image](https://github.com/user-attachments/assets/a7edd2fb-91ff-4f9b-b5eb-04b14ee7b7a2)

I also moved the date parsing to `instructorCourseAdminInstances.ts`, within the POST request function. There is also validation here to ensure that end_access_date comes after start_access_date.

![image](https://github.com/user-attachments/assets/d3ee75ea-3098-40a5-8829-872232fc6a9d)